Ubase64 bugfix

createyourpersonalaccount opened the pull request:
<!-- Filling this template is mandatory -->

**Your checklist for this pull request**
- [X] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [X] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [X] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [X] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

<!-- Explain the **details** for making this change. Is a new feature implemented? What existing problem does the pull request solve? How does the pull request solve these issues? Please provide enough information so that others can review your pull request. -->

The original Base64 code had some bugs which are demonstrated by the unit tests included in the first commit. The second commit fixes the issues.

**Test plan**

<!-- What steps should the reviewer take to test your pull request? Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots/videos. This is your time to re-check that everything works and that you covered all the edge cases -->

...

**Closing issues**

<!-- put "closes #XXXX" in your comment to auto-close the issue that your PR fixes (if any). -->

...

createyourpersonalaccount committed:
test: some base64 unit tests to demonstrate bugs

The Base64 implementation has some bugs in its decoding function that
we demonstrate with unit tests:

- In test_rz_base64_decode_invalid, we add a unit test to show that
the invalid characters are not ignored correctly. I believe the
intention of the original code is to ignore them, however this is not
correctly achieved. The two common options, ignoring invalid
characters or rejecting the message, are mentioned in RFC 4648. What
the current base64 decoding function does is a complicated logic that
may reject, or otherwise, it appears to overwrite previously encoded
data.

We show this in the test by decoding two inequivalent Base64-encoded
messages, which under current implementation both result in the same
string "foo", but they should be different.

- In test_rz_base64_decode_offby1, we demonstrate an off-by-1 buffer
overflow. An initial buffer of content { 'A', 0, 'B', 0 } is passed to
the encoding function with intention of encoding the single character
'A'. When decoding back into plaintext, the decoding function is:

1) inserting a NUL byte (out of courtesy?) which may not be
desirable. However, this NUL byte is already present in our original
buffer.
2) Reaching past the NUL byte, overwriting the character 'B' with a
NUL byte.

We demonstrate 2) with the assertion in the unit test.

createyourpersonalaccount committed:
librz: rewrite base64 code

The new Base64 code is fixing the previous bugs which were illustrated
with unit tests. Documentation is added and as a "breaking" API
change, the input and output parameters are allowed to be NULL
consistently (not just some, but all.) The rest of the behavior is the
same, save for the bugs that were fixed, and the decoding function
which is now even laxer in what it is able to decode (which seemed to
be the original intention but had a bug as demonstrated by a unit
test.)

An integer overflow bug, in the calculation of the string length, is
also fixed.

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+ * If the parameter \p src is \c NULL, nothing is done and the value
+ * \c NULL is returned.
+ */
+RZ_API ut8 *rz_base64_decode_dyn(const char *src, int len) {
```
tbh i'm not happy to have `len` as `int`. probably better as `st64`

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+	ret_size += 1; // include NUL byte
+	// we attempt to minimize memory usage
+	if ((tmp = realloc(ret, ret_size)) != NULL) {
+		return tmp;
+	} else {
+		return ret;
+	}
```
remove the else. just return.

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+	}
+	ret_size = 1 + 3 * (len / 4);
+	ret = malloc((size_t)ret_size);
+	if (ret == NULL) {
```
```suggestion
	if (!ret) {
```

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+		dest[0] = tmp[0];
+		dest[1] = '\0';
+		return ret + 1;
+	} else /* j == 3 */ {
+		buf[3] = 0;
+		unpack_from6((ut8 *)tmp, (const ut8 *)buf);
+		dest[0] = tmp[0];
+		dest[1] = tmp[1];
+		dest[2] = '\0';
+		return ret + 2;
 	}
```
remove the else

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+static void unpack_from6(ut8 dest[3], const ut8 src[4]) {
+	ut8 idx[4];
+	size_t i;
+	for (i = 0; i < 4; i++)
+		idx[i] = cd64(src[i]);
```
add parenthesis

createyourpersonalaccount replied:
Where?

createyourpersonalaccount replied:
You mean brackets, okay.

wargio replied:
yes sorry.

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,351 @@
+		return 62;
+	} else if (c == '/') {
+		return 63;
+	} else {
+		return SIZE_MAX;
```
remove else.

wargio reviewed (changes requested):
add your name in the copyright

createyourpersonalaccount commented:
> add your name in the copyright

How specifically should I do this?

edit: done

createyourpersonalaccount committed:
librz: address PR review issues

1. Add name to copyright.
2. Modify public API type from int to st64.

For style reasons, also:

1. Remove 'else' statements.
2. Add brackets.
3. Prefer ! operator to NULL comparison.

createyourpersonalaccount committed:
librz: update documentation and fix overflow bug

We update the documentation with examples and fix an overflow bug in
the calculation of the decoded message. The documentation for this
decoded message length is also updated.

wargio commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,418 @@
 /* Original code from:
  * dmc - dynamic mail client -- author: pancake
  * See LICENSE file for copyright and license details.
+ *
+ * Updated in 2023 by Nikolaos Chatzikonstantinou
+ * <nchatz314@gmail.com>.
```
no, just add it at the top as `// SPDX-FileCopyrightText: 2023 Nikolaos Chatzikonstantinou <nchatz314@gmail.co>`

createyourpersonalaccount replied:
Just add this line without modifying the previous `SPDX-FileCopyrightText`?

XVilka replied:
Yes, just SPDX line is enough, remove this "Updated..."

wargio reviewed (approved):
LGTM, just fix the copyright info.

createyourpersonalaccount committed:
librz: add copyright header and change char to ut8 in example

XVilka commented on librz/include/rz_util/rz_base64.h:
```diff
@@ -8,8 +8,8 @@ extern "C" {
-RZ_API int rz_base64_decode(ut8 *bout, const char *bin, int len);
-RZ_API ut8 *rz_base64_decode_dyn(const char *in, int len);
+RZ_API st64 rz_base64_decode(ut8 *bout, const char *bin, st64 len);
+RZ_API ut8 *rz_base64_decode_dyn(const char *in, st64 len);
```
While you are on it, could you please also add ownership marks? Like described in DEVELOPERS.md

XVilka commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,419 @@
+ * free(bin_dec);
+ * \endcode
+ */
+RZ_API char *rz_base64_encode_dyn(const ut8 *src, size_t n) {
```
Ownership

createyourpersonalaccount replied:
What do you mean by ownership?

XVilka commented on librz/util/ubase64.c:
```diff
@@ -1,104 +1,419 @@
+ *
+ * See \a rz_base64_encode_dyn for an example.
+ */
+RZ_API ut8 *rz_base64_decode_dyn(const char *src, st64 len) {
```
Ownership

createyourpersonalaccount committed:
librz: add ownership marks and update copyright

wargio reviewed (approved):
lgtm

wargio commented:
please make this green.
```
Mismatched function annotation for rz_base64_encode_dyn at <rizin/librz/include/rz_util/rz_base64.h:13:14> : None (was RZ_OWN at <rizin/librz/util/ubase64.c:253:21>
Mismatched function annotation for rz_base64_decode_dyn at <rizin/librz/include/rz_util/rz_base64.h:12:13> : None (was RZ_OWN at <rizin/librz/util/ubase64.c:392:20>
E
```

createyourpersonalaccount commented:
@wargio thanks I will.

createyourpersonalaccount committed:
librz: fix type signature in header to include ownership

wargio commented on librz/include/rz_util/rz_base64.h:
```diff
@@ -8,9 +8,9 @@ extern "C" {
-RZ_API int rz_base64_decode(ut8 *bout, const char *bin, int len);
-RZ_API ut8 *rz_base64_decode_dyn(const char *in, int len);
-RZ_API char *rz_base64_encode_dyn(const ut8 *bin, size_t sz);
+RZ_API st64 rz_base64_decode(ut8 *bout, const char *bin, st64 len);
+RZ_API RZ_OWN ut8 *rz_base64_decode_dyn(const char *in, st64 len);
+RZ_API RZ_OWN char *rz_base64_encode_dyn(const ut8 *bin, size_t sz);
```
please add also `RZ_NONNULL` to the function arguments which accepts pointers and for `bout` also `RZ_OUT` (don't forget to update also the .c function definitions).

createyourpersonalaccount replied:
Does `RZ_NONNULL` work together with `rz_return_val_if_fail(src, NULL)`? Because technically, you could pass `NULL`, it's even documented in the docstring.

wargio replied:
yes, if null is accepted, then you should use `RZ_NULLABLE`

wargio reviewed (changes requested):
please add these changes and we can merge. awesome work btw.

createyourpersonalaccount committed:
librz: add nullable pointer modifier
